### PR TITLE
feat: add native training snapshot support for deterministic resume

### PR DESCRIPTION
**Pull request by aristotelisxs**
## Summary

This PR adds native training snapshot / checkpoint support to LightGBM, allowing interrupted training to be resumed from a saved snapshot.

The implementation is integrated across the C++ core, CLI, C API, and Python package, with deterministic resume behavior validated against uninterrupted training runs.

## What changed

- added versioned binary snapshot serialization in the core training loop
- added atomic snapshot writes using temp-file replacement
- saved and restored training-critical state, including:
  - model trees
  - training iteration state
  - score buffers
  - bagging / sampling state
  - relevant RNG state
  - tree learner state
  - DART-specific state
- added C API support for:
  - saving snapshots
  - loading snapshots
  - querying current training iteration
- added Python support for:
  - `save_snapshot`
  - `snapshot_freq`
  - `snapshot_path`
  - auto-resume when `snapshot_path` already exists
- added CLI support for snapshot save / resume
- updated parameter documentation

## Behavior

Default behavior is unchanged unless snapshot-related parameters are set.

In Python, snapshot support is enabled with parameters like:

```python
params = {
    "objective": "binary",
    "save_snapshot": True,
    "snapshot_freq": 10,
    "snapshot_path": "model.snapshot",
}

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for your interest in LightGBM.

Are there some prior discussions (here or in other projects) motivating this?

LightGBM already supports training continuation via loading from model text files or in-memory Boosters... what does this add that the existing training-continuation support in LightGBM is missing?

This is a huge amount of new code and maintenance burden to propose at once, with 0 prior discussion and when you have (as far as I know) never interacted with the project before. Unfortunately the wide proliferation of LLM coding tools has made it cheap to generate such PRs, so I'm skeptical by default of such unexpected contributions. LightGBM is maintained by a small group of busy volunteers, and we have to be very protective of our limited time and attention.

If you can answer these questions and demonstrate a willingness to work with us on making this align with LightGBM's goals, we'll be happy to consider this. If not, this PR will be closed.

**Comment by aristotelisxs** [comment]
Thanks for the thoughtful response, and I appreciate the maintainers’ time and constraints.

You’re absolutely right that LightGBM already supports training continuation via `init_model`. The motivation here is not to replace that, but to address automatic fault tolerance during long training runs. This has been a personal struggle, as I've used LightGBM extensively in previous projects.

Today, continuation in LightGBM requires users to explicitly implement checkpointing logic, for example via callbacks or periodic saves. If a process crashes before such a save, all intermediate progress since the last checkpoint is lost.

The proposed feature is closer to built-in periodic snapshotting, where:
- training state is saved automatically at intervals during `fit()`
- on restart, training resumes from the latest snapshot with no additional user code

So the gap is not capability, but out-of-the-box crash recovery with minimal user intervention, which is especially relevant for long GPU jobs.

Regarding prior discussion, that is a fair point. I should have opened an issue first to validate the idea and scope before putting together a larger PR. Happy to step back and do that if preferred.

I’m also very open to aligning this with LightGBM’s design goals, or scoping it down significantly if a lighter approach makes more sense. For example, even a minimal callback-based reference implementation or discussion around best practices for checkpointing could already provide value.

Especially on the C API side of the implementation, I have used Codex to help me scope and implement changes/additions, as I do not have much experience there. I understand this has over-inflated the codebase. I will revise this as part of next steps.

Let me know how you would prefer to proceed, and I’m happy to iterate accordingly.

**Comment by aristotelisxs** [comment]
Given the sheer size of this PR, I introduced some refinements to reduce it with the latest commit. 

Some known limitations I identified FYI:

- Snapshot save / resume is not supported for distributed training and ranking objectives (`lambdarank`, `rank_xendcg`).
- `save_snapshot=True` is not compatible with the `early_stopping()` callback.
- Resume will fail if the snapshot file is missing, corrupted, or incompatible with the current training setup.
- Resume is only supported for the currently wired local tree learner and sampling strategy implementations.

**Comment by jameslamb** [comment]
Thanks for the response and your interest in LightGBM, but I'm closing this PR.

> _I have used Codex to help me scope and implement changes/additions_

Your more recent commit (https://github.com/lightgbm-org/LightGBM/pull/7226/changes/103e38d28c7eb18553840e176e63fc9f15f97e71) contains totally-unrelated changes like updating all of the powershell linting code... that's evidence to me that you're not reviewing your own changes before pushing them here.

image

As another example, I don't understand why we need a new API `LGBM_BoosterGetCurrentTrainingIteration()` when LightGBM already has a `LGBM_BoosterGetCurrentIteration()`. That's exactly the type of complexity that LLMs tend to unnecessarily add unless carefully instructed. We're also very protective of the C API. It's a public ABI / API we promise to maintain and not break without very very good reasons. I'm not supportive of this PR expanding it this way, this much, without a much clearer statement of the benefit.

Using LLMs is fine, but if you want us to merge and maintain all this code, you should review and understand it yourself. We are tired and busy people, and have to choose where to spend our limited time... I think this PR is far away from being something we'll merge, and I'm not willing to invest the time needed to review it any further.

> _even a minimal callback-based reference implementation or discussion around best practices for checkpointing could already provide value_
>
> _I should have opened an issue first to validate the idea and scope before putting together a larger PR. Happy to step back and do that if preferred_

You're welcome to open a feature request issue, describing the functionality you want and why LightGBM needs to change to support it. Examples from other similar libraries with similar functionality would help make your case. Maybe some other maintainer or other LightGBM users will take interest in it and we can go from there, ideally with a very limited.

If you could achieve something like this with only callbacks, that'd a be a very interesting demonstration of the desired behavior! And probably a good starting point for anyone else wanting to write self-healing training code like what you're referring to.

> _training state is saved automatically at intervals during `fit()`_
> _on restart, training resumes from the latest snapshot with no additional user code_

I'm not totally convinced that LightGBM needs to automatically support this via `params` and new functions. The Python package already has APIs you could use to write this. Maybe we could start by making that pattern work?

For example, I'd expected it'd be possible to write a neural-network-style training loop with something like this:

```python
from sklearn.datasets import make_regression
import lightgbm as lgb
import os

NUM_BOOST_ROUNDS = 15

X, y = make_regression(n_samples=10_000, n_features=5, n_informative=5, random_state=312)

snapshot_file = "model.txt"

model_file = None
if os.path.exists(snapshot_file):
    model_file = snapshot_file

bst = lgb.Booster(
    params={
        "objective": "regression",
        "snapshot_freq": 1,
    },
    train_set = lgb.Dataset(X, label=y),
    model_file=snapshot_file,
)

while bst.current_iteration() < NUM_BOOST_ROUNDS:
    print(f"training for iter: {bst.current_iteration()}")
    bst.update()
    bst.save_model(snapshot_file)

print(f"done training. Model saved at '{snapshot_file}")
```

From my brief experiments tonight that seems not to quite work (`.num_trees()` returns `0` when creating a `Booster` from a model file, it seems).

Maybe identifying a workflow similar to that that we could add in the `examples/` directory, plus fixing bugs found along the way to support it, would be a better start on this problem.
